src/foam/net/node/Server.js

```diff
@@ -0,0 +1,125 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'foam.net.node',
+  name: 'Server',
+
+  requires: [
+    'foam.dao.ArrayDAO',
+    'foam.node.handlers.FileHandler',
+    'foam.node.handlers.RestDAOHandler',
+    'foam.node.handlers.StaticFileHandler'
+  ],
+
+  imports: [
+    'info',
+    'log'
+  ],
+
+  properties: [
+    {
+      class: 'FObjectArray',
+      of: 'foam.net.node.Handler',
+      name: 'handlers'
+    },
+    {
+      type: 'Int',
+      name: 'port',
+      value: 8000
+    },
+    {
+      name: 'server',
+      documentation: 'The Node JS HTTP Server object.',
+      value: null
+    },
+    {
+      name: 'http',
+      factory: function() { return require('http'); }
+    }
+  ],
+  methods: [
+    function start() {
+      if ( this.server ) return Promise.resolve(this.server);
+
+      this.server = this.http.createServer(this.onRequest);
+
+      var self = this;
+      return new Promise(function(resolve, reject) {
+        self.server.listen(self.port, function() {
+        self.info(
+          self.handlers.length + ' handlers listening on port ' + self.port);
+          resolve(self.server);
+        });
+      });
+    },
+    function shutdown() {
+      if ( ! this.server ) return Promise.resolve(null);
+
+      var self = this;
+      return new Promise(function(resolve, reject) {
+        self.server.close(function() {
+          self.server = null;
+          resolve(null);
+        });
+      });
+    },
+    function addHandler(handler) {
+      // TODO(adamvy): Not wild about this design, consider a better model for handling
```

Comment:
No. In fact, I'm happy to let the server register a flat list of handlers and support handler decoration (orthogonal to the server's logic) for things like OAuth.